### ISSUE TITLE
Getting a 400 {object object} when using this in my jasmine tests

### ISSUE THREAD

**Issue by jordandlaman** [mvndaai/testrail-promise#13]
Here is my code from a protractor test

```import {element, browser, by } from 'protractor';
import { TestRail } from 'testrail-promise';

const resultLeaker = {
  suiteStarted: function(result){ (jasmine as any).results = {suite:result}; },
  specStarted: function(result){ (jasmine as any).results.spec = result; }
};
jasmine.getEnv().addReporter(resultLeaker);
describe('BLAH of concept protractor to testrail integration through api', () => {
  var TestRail = require("testrail-promise");
  var tr = new TestRail("websiteurigohere", "usernamegoeshere", "apigoeshere");
      it('Passing', function(){ expect(true).toBeTruthy(); });
      it('Failing', function(){ expect(false).toBeTruthy(); });
  afterEach(function(done){
          reportToTestRail(this).finally(function(){ done(); });
      });
      var reportToTestRail = function(test){
          var obj = {
              "project_name":"Clinical Integration Services",
              "plan_name":"AA",
              "section_name":"Test Section",
              "title":(jasmine as any).results.spec.fullName,
              "status_name":((jasmine as any).results.spec.failedExpectations.length === 0 ? "passed" : "failed"),
              "add_results":((jasmine as any).results.spec.failedExpectations.length === 0 ? "passed" : "failed"),
            };
            tr.simpleRequests();
          return tr.ifNeededCreateThenAddResultForCase(obj);
      };
    });
```

Here is the error when using simpleRequests

```
Unhandled rejection StatusCodeError: 400 - [object Object]
    at new StatusCodeError (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/errors.js:26:15)
    at Request.RP$callback [as _callback] (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/rp.js:68:32)
    at Request.self.callback (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1091:12)
    at IncomingMessage.g (events.js:292:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
.Unhandled rejection StatusCodeError: 400 - [object Object]
    at new StatusCodeError (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/errors.js:26:15)
    at Request.RP$callback [as _callback] (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/rp.js:68:32)
    at Request.self.callback (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1091:12)
    at IncomingMessage.g (events.js:292:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
```

I am successfully adding a testcase and testresult into testrails but there are two problems.  The first and biggest one is that no matter what I do the test results will never have the status set to anything but untested.  The second is that this plugin is sticking the testresults into the latest test run.  I am not sure how to specify where to put them.  Thanks in advance.

**Comment by mvndaai** [comment]
I am not sure what issue you are having to make to so the results never get set. You can try going step by step through the function to figure out where it is going wrong https://github.com/mvndaai/testrail-promise/blob/master/lib/helpers.js#L146

About your uncaught error try adding a catch  to this line `reportToTestRail(this).finally(function(){ done(); });`
make it something like this
```
reportToTestRail(this)
.catch((a) => console.log(a))
.finally(function(){ done(); });
```
It looks like the function `ifNeededCreateThenAddResultForCase` doesn't have an option to not grab the latest run if one is provided If you want to put in a pull request for that I would love to push it.

**Comment by mvndaai** [comment]
This should let you add your run id https://github.com/mvndaai/testrail-promise/commit/20c8311f99007a551a722143e0448a7227a325d2

**Comment by jordandlaman** [comment]
Let me just say thank you so much for getting back to this issue and trying to rectify a problem so quickly.  I will implement that commit right away.  Using the catch I get 
```[13:20:42] I/launcher - Running 1 instances of WebDriver
[13:20:42] I/hosted - Using the selenium server at http://localhost:4444/wd/hub
Started
{ StatusCodeError: 400 - [object Object]
    at new StatusCodeError (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/errors.js:26:15)
    at Request.RP$callback [as _callback] (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request-promise/lib/rp.js:68:32)
    at Request.self.callback (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (/Users/m54148/sites/protractor/exampleTypescript/node_modules/request/request.js:1091:12)
    at IncomingMessage.g (events.js:292:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
  name: 'StatusCodeError',
  statusCode: 400,
  message: '400 - [object Object]',
  error: { error: 'No (active) test found for the run/case combination.' },
  options:
   { simple: true,
     uri: 'http://cilappxs0021.sys.cigna.com/testrail/index.php?/api/v2/add_result_for_case/1454/6736957',
     json: true,
     rejectUnauthorized: false,
     headers: { 'content-type': 'application/json' },
     body:
      { project_name: 'Clinical Integration Services',
        plan_name: 'AA',
        section_name: 'Test Section',
        title: 'pass of concept protractor to testrail integration through api Passing',
        status_id: 1,
        status_name: 'passed',
        estimate: '',
        refs: '',
        type_id: 3,
        suite_name: 'Master',
        project_id: 11,
        suite_id: 11,
        plan_id: 1374,
        section_id: 410195,
        run_id: 1454,
        case_id: 6736957 },
     method: 'POST',
     callback: undefined,
     transform: undefined,
     resolveWithFullResponse: false },
  response:
   IncomingMessage {
     _readableState:
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: [Object],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: true,
        endEmitted: true,
        reading: false,
        sync: true,
        needReadable: false,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: false,
     domain: null,
     _events:
      { end: [Object],
        close: [Object],
        data: [Function],
        error: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     socket:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'cilappxs0021.sys.cigna.com',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        _bytesDispatched: 689,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     connection:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'cilappxs0021.sys.cigna.com',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        _bytesDispatched: 689,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     httpVersion: '1.1',
     complete: true,
     headers:
      { date: 'Fri, 22 Sep 2017 20:20:47 GMT',
        server: 'Apache/2.2.15 (Red Hat)',
        'x-powered-by': 'PHP/5.3.3',
        'content-length': '65',
        connection: 'close',
        'content-type': 'application/json; charset=utf-8',
        'x-rbt-optimized-by': 'ORGA-A-05A-WO-1A.internal.cigna.com (RiOS 9.1.3a) IK' },
     rawHeaders:
      [ 'Date',
        'Fri, 22 Sep 2017 20:20:47 GMT',
        'Server',
        'Apache/2.2.15 (Red Hat)',
        'X-Powered-By',
        'PHP/5.3.3',
        'Content-Length',
        '65',
        'Connection',
        'close',
        'Content-Type',
        'application/json; charset=utf-8',
        'X-RBT-Optimized-By',
        'ORGA-A-05A-WO-1A.internal.cigna.com (RiOS 9.1.3a) IK' ],
     trailers: {},
     rawTrailers: [],
     upgrade: false,
     url: '',
     method: null,
     statusCode: 400,
     statusMessage: 'Bad Request',
     client:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: null,
        _parent: null,
        _host: 'cilappxs0021.sys.cigna.com',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        _bytesDispatched: 689,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     _consuming: true,
     _dumped: false,
     req:
      ClientRequest {
        domain: null,
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        output: [],
        outputEncodings: [],
        outputCallbacks: [],
        outputSize: 0,
        writable: true,
        _last: true,
        upgrading: false,
        chunkedEncoding: false,
        shouldKeepAlive: false,
        useChunkedEncodingByDefault: true,
        sendDate: false,
        _removedHeader: [Object],
        _contentLength: null,
        _hasBody: true,
        _trailer: '',
        finished: true,
        _headerSent: true,
        socket: [Object],
        connection: [Object],
        _header: 'POST /testrail/index.php?/api/v2/add_result_for_case/1454/6736957 HTTP/1.1\r\ncontent-type: application/json\r\nhost: cilappxs0021.sys.cigna.com\r\naccept: application/json\r\ncontent-length: 368\r\nauthorization: Basic am9yZGFuLmxhbWFuQGNpZ25hLmNvbTpYS2N6YndMbmFzeUFUS0wvMUQ2Ty1uY29DME1NcXhMczFGVXp4eW5RaQ==\r\nConnection: close\r\n\r\n',
        _headers: [Object],
        _headerNames: [Object],
        _onPendingData: null,
        agent: [Object],
        socketPath: undefined,
        timeout: undefined,
        method: 'POST',
        path: '/testrail/index.php?/api/v2/add_result_for_case/1454/6736957',
        _ended: true,
        parser: null,
        res: [Circular] },
     request:
      Request {
        domain: null,
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        simple: true,
        uri: [Object],
        rejectUnauthorized: false,
        headers: [Object],
        body: '{"project_name":"Clinical Integration Services","plan_name":"AA","section_name":"Test Section","title":"pass of concept protractor to testrail integration through api Passing","status_id":1,"status_name":"passed","estimate":"","refs":"","type_id":3,"suite_name":"Master","project_id":11,"suite_id":11,"plan_id":1374,"section_id":410195,**"run_id":1454**,"case_id":6736957}',
        method: 'POST',
        readable: true,
        writable: true,
        explicitMethod: true,
        _qs: [Object],
        _auth: [Object],
        _oauth: [Object],
        _multipart: [Object],
        _redirect: [Object],
        _tunnel: [Object],
        _rp_resolve: [Function],
        _rp_reject: [Function],
        _rp_promise: [Object],
        _rp_callbackOrig: undefined,
        callback: [Function],
        _rp_options: [Object],
        setHeader: [Function],
        hasHeader: [Function],
        getHeader: [Function],
        removeHeader: [Function],
        localAddress: undefined,
        pool: [Object],
        dests: [],
        __isRequestRequest: true,
        _callback: [Function: RP$callback],
        proxy: null,
        tunnel: false,
        setHost: true,
        originalCookieHeader: undefined,
        _disableCookies: true,
        _jar: undefined,
        port: 80,
        host: 'cilappxs0021.sys.cigna.com',
        path: '/testrail/index.php?/api/v2/add_result_for_case/1454/6736957',
        _json: true,
        httpModule: [Object],
        agentClass: [Object],
        agent: [Object],
        _started: true,
        href: 'http://cilappxs0021.sys.cigna.com/testrail/index.php?/api/v2/add_result_for_case/1454/6736957',
        req: [Object],
        ntick: true,
        response: [Circular],
        originalHost: 'cilappxs0021.sys.cigna.com',
        originalHostHeaderName: 'host',
        responseContent: [Circular],
        _destdata: true,
        _ended: true,
        _callbackCalled: true },
     toJSON: [Function: responseToJSON],
     caseless: Caseless { dict: [Object] },
     read: [Function],
     body: { error: 'No (active) test found for the run/case combination.' } } }```

Investigating this output it seems like that last response is the problem.  It can't find the case/run combination to update it.  When I parsed through what it is sending the api in terms of run id and check that with where it is actually putting it there is an inconsistency.  It wants to look for the correct case_id but the run_id is not where the test case is ending up.  It should be looking for run_id 1444.  Not sure why that it is though.

**Comment by mvndaai** [comment]
On one line it has 
```
body: '{"project_name":"Clinical Integration Services","plan_name":"AA","section_name":"Test Section","title":"pass of concept protractor to testrail integration through api Passing","status_id":1,"status_name":"passed","estimate":"","refs":"","type_id":3,"suite_name":"Master","project_id":11,"suite_id":11,"plan_id":1374,"section_id":410195,**"run_id":1454**,"case_id":6736957}',
```
It has `**"run_id":1454**`. Does run 1454 exist in your testrail?

**Comment by jordandlaman** [comment]
Yeah it was placing the tests in a run but looking for them in a completely different run.  Thanks again for helping to diagnose this.  I actually talked to the guys managing our testrail instance and I think it may have had to do with using the string matching of the library and our file structure.  There were identical folders for different teams in section and plan and thus I think it was getting a bit confused where to put stuff and where to find it.  I switched to a different folder structure with unique names and your library had no problem doing its job after I told it what run_id to look for.  Thanks so much.
